docs/src/cookbooks/serialization.md

```diff
@@ -0,0 +1,177 @@
+Thus the `SerializableModuleGenerator`, `SerializableModule[T <: SerializableModuleParameter]` and `SerializableModuleParameter` is provided to solve this issue.
+For any `SerializableModuleGenerator`, Chisel can automatically serialize and de-serialize it by adding a constraints:
+1. the `SerializableModule` should not be inner class, since the outer class is parameter to it;
+1. the `SerializableModule` has and only has one parameter with `SerializableModuleParameter` as its type.
+1. the Module cannot depends on global variables and use non-reproducible function(random, time, etc), and this should be guaranteed by user, since Scala cannot detect it.
+
+It can provide these benefits:
+1. user can use `SerializableModuleGenerator(module: class[SerializableModule], parameter: SerializableModuleParameter)` to auto serialize a Module and its parameter.
+1. user can nest `SerializableModuleGenerator` in other serializable parameters to represent a relative large parameter.
+1. user can elaborate any `SerializableModuleGenerator` into a single module for testing.
+
+
+## How do I serialize Modules with `SerializableModuleGenerator`
+It is pretty simple and illustrated by example below, the GCD Module with width as its parameter.
+
+```scala mdoc:silent
+import chisel3._
+import chisel3.experimental.{SerializableModule, SerializableModuleGenerator, SerializableModuleParameter}
+import upickle.default._
+
+// provide serialization functions to GCDSerializableModuleParameter
+object GCDSerializableModuleParameter {
+  implicit def rwP: ReadWriter[GCDSerializableModuleParameter] = macroRW
+}
+
+// Parameter
+case class GCDSerializableModuleParameter(width: Int) extends SerializableModuleParameter
+
+// Module
+class GCDSerializableModule(val parameter: GCDSerializableModuleParameter)
+    extends Module
+    with SerializableModule[GCDSerializableModuleParameter] {
+  val io = IO(new Bundle {
+    val a = Input(UInt(parameter.width.W))
+    val b = Input(UInt(parameter.width.W))
+    val e = Input(Bool())
+    val z = Output(UInt(parameter.width.W))
+  })
+  val x = Reg(UInt(parameter.width.W))
+  val y = Reg(UInt(parameter.width.W))
+  val z = Reg(UInt(parameter.width.W))
+  val e = Reg(Bool())
+  when(e) {
+    x := io.a
+    y := io.b
+    z := 0.U
+  }
+  when(x =/= y) {
+    when(x > y) {
+      x := x - y
+    }.otherwise {
+      y := y - x
+    }
+  }.otherwise {
+    z := x
+  }
+  io.z := z
+}
+```
+using `write` function in `upickle`:
+```scala
+val j = upickle.default.write(
+  SerializableModuleGenerator(
+    classOf[GCDSerializableModule],
+    GCDSerializableModuleParameter(32)
+  )
+)
+println(j)
```

Comment:
````suggestion
```scala mdoc:silent
val j = upickle.default.write(
  SerializableModuleGenerator(
    classOf[GCDSerializableModule],
    GCDSerializableModuleParameter(32)
  )
)
````

Is the constraint on no inner modules only at runtime or also at compile time? If you can still compile it you could do `mdoc:silent` (also please remove the println, especially becuase it doesn't execute).

Reply:
It's runtime. I just took a look at the doc of mdoc, thanks Jack.

Reply:
I restore the silent in json dump command, since it's a one-line result.